tests/components/broadlink/test_device.py

```diff
@@ -0,0 +1,389 @@
+
+    assert mock_entry.state == ENTRY_STATE_NOT_LOADED
+    forward_entries = {c[1][1] for c in mock_forward.mock_calls}
+    domains = get_domains(mock_api.type)
+    assert mock_forward.call_count == len(domains)
+    assert forward_entries == domains
+
+
+async def test_device_unload_authentication_error(hass):
+    """Test we unload a device that failed the authentication step."""
+    device = get_device("Living Room")
+    mock_api = device.get_mock_api()
+    mock_api.auth.side_effect = blke.AuthenticationError()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ), patch.object(hass.config_entries.flow, "async_init"):
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    with patch.object(
+        hass.config_entries, "async_forward_entry_unload", return_value=True
+    ) as mock_forward:
+        await hass.config_entries.async_unload(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_NOT_LOADED
+    assert mock_forward.call_count == 0
+
+
+async def test_device_unload_update_failed(hass):
+    """Test we unload a device that failed the update step."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_api.check_sensors.side_effect = blke.DeviceOfflineError()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ):
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    with patch.object(
+        hass.config_entries, "async_forward_entry_unload", return_value=True
+    ) as mock_forward:
+        await hass.config_entries.async_unload(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_NOT_LOADED
+    assert mock_forward.call_count == 0
+
+
+async def test_device_update_listener(hass):
+    """Test we update device and entity registry when the entry is renamed."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    device_registry = mock_device_registry(hass)
+    entity_registry = mock_registry(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api):
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+        await hass.async_block_till_done()
+
+        hass.config_entries.async_update_entry(mock_entry, title="New Name")
+        await hass.async_block_till_done()
+
+    device_entry = device_registry.async_get_device(
+        {(DOMAIN, mock_entry.unique_id)}, set()
+    )
+    assert device_entry.name == "New Name"
+    for entry in async_entries_for_device(entity_registry, device_entry.id):
+        assert entry.original_name.startswith("New Name")
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6042836)